### PR TITLE
Create wide table

### PR DESCRIPTION
**Pull request by dfalster**
Draft function to create wide table

Started by @dcdcol2804, expanded by @dfalster, input from @ehwenk 

```
remotes::install_github("traitecoevo/austraits@big_table")
```
or checkout branch
```
dev_tools::load_all
```

Then try run with
```
austraits <- load_austraits("3.0.2", path = "ignore/data/austraits/")
austraits_wide <- austraits %>% as_wide_table()
```

Preserves number of rows in trait dataset

```
> austraits$traits %>% nrow()
[1] 997808
> austraits_wide %>% nrow()
[1] 997808
```

This version is much faster too. On my laptop, previous version took 55s, this one takes 14s.

 Output is as follows
```
austraits_wide
# A tibble: 997,808 × 31
   dataset_id  observation_id  trait_name  taxon_name trait_value unit  value_type replicates
   <chr>       <chr>           <chr>       <chr>      <chr>       <chr> <fct>      <chr>     
 1 Ahrens_2019 Ahrens_2019_001 leaf_area   Corymbia … 12883.7     mm2   raw_value  1         
 2 Ahrens_2019 Ahrens_2019_001 leaf_delta… Corymbia … -27.98      per … raw_value  1         
 3 Ahrens_2019 Ahrens_2019_001 leaf_N_per… Corymbia … 9           mg/g  raw_value  1         
 4 Ahrens_2019 Ahrens_2019_001 leaf_PRI    Corymbia … 0.022642696 umol… raw_value  1         
 5 Ahrens_2019 Ahrens_2019_001 modified_N… Corymbia … 0.711535497 dime… raw_value  1         
 6 Ahrens_2019 Ahrens_2019_001 specific_l… Corymbia … 5.4         mm2/… raw_value  1         
 7 Ahrens_2019 Ahrens_2019_001 water_band… Corymbia … 1.060028751 dime… raw_value  1         
 8 Ahrens_2019 Ahrens_2019_001 wood_densi… Corymbia … 0.600120627 mg/m… raw_value  1         
 9 Ahrens_2019 Ahrens_2019_002 leaf_area   Corymbia … 9647.7      mm2   raw_value  1         
10 Ahrens_2019 Ahrens_2019_002 leaf_delta… Corymbia … -27.33      per … raw_value  1         
# … with 997,798 more rows, and 23 more variables: date <chr>, collection_type <chr>,
#   sample_age_class <chr>, sampling_strategy <chr>, latitude (deg) <chr>,
#   longitude (deg) <chr>, site_name <chr>, site <chr>, context_name <chr>, context <chr>,
#   methods <chr>, original_name <chr>, dataset_description <chr>,
#   source_primary_citation <chr>, source_secondary_citation <chr>, taxonomicStatus <chr>,
#   taxonDistribution <chr>, taxonRank <chr>, genus <chr>, family <chr>,
#   acceptedNameUsageID <chr>, scientificNameAuthorship <chr>, ccAttributionIRI <chr>

```

### CODE REVIEW

File: R/as_big_trait_table.R [hunk 1]
```diff
@@ -1,119 +1,133 @@
...
-  # we only need two extra columns from the trait class table - collapsing two category and other_tags cols and renaming them for clarity
-  x2 = trait_class %>% mutate(trait_category = str_c(category, "; ", other_tags)) %>% 
-    select(trait_name, tissue, trait_category)
-  
-  x2$trait_category = gsub("; $", "", x2$trait_category)
+  austraits$traits <- 
+    austraits$traits %>% 
+    dplyr::rename(c("trait_value" = "value")) 
   
   # The contexts table needs the contexts collapsed to one context name per site
```
Comment (ehwenk) [inline R/as_big_trait_table.R]: What does it mean that we need "one context name per site"? The number of contexts allowed should be independent of the site name. There are often multiple contexts per site.

File: R/as_big_trait_table.R [hunk 2]
```diff
@@ -1,119 +1,133 @@
...
-  x5 = merge(sites_lat_long, sites_condensed, by = c("dataset_id", "site_name"))
-  
-  #remove the unwanted extra NA text from the collapse_cols function
-  x5$a = strsplit(x5$site,";")
-  x5$b = lapply(x5$a, function(x){ subset(x, grepl("'NA'", x) == F)})
-  x5$site = unlist(lapply(x5$b, function(x){ str_c(x, collapse = "; ")}))
-  x5 = x5 %>% select(-a, -b)
-  
+    dplyr::filter(value!="unkown") %>% 
+    # next line is a fix -- one dataset in 3.0.2 has value "site_name"
```
Comment (ehwenk) [inline R/as_big_trait_table.R]: Did you mean to leave out the line that filters out `latitude` and `longitude`? ```filter(site_property != "latitude (deg)" & site_property != "longitude (deg)")```

File: R/as_big_trait_table.R [hunk 3]
```diff
@@ -1,119 +1,133 @@
...
     # The most useful (if you are filtering for just one taxon_name)
-    dataset_id, trait_name, trait_value, unit, 
-    value_type, replicates, tissue, trait_category,
+    dataset_id, observation_id, trait_name, taxon_name, trait_value, unit, 
+    value_type, replicates, 
+    # tissue, trait_category,  # Add after new zenodo release
     
     # More stuff you can filter on
-    collection_type, sample_age_class, sampling_strategy, observation_id, date,
+    date, collection_type, sample_age_class, sampling_strategy,
```
Comment (ehwenk) [inline R/as_big_trait_table.R]: just to flag, `data` will become `collection_date`

**Review by ehwenk — APPROVED** [review]
I've read through the code and run through all the components and looked at the output and everything looks good.

**Comment by fontikar** [comment]
How cool!! Well done folks! Let me test it out and write a test for it before merging in! 

> This version is much faster too. On my laptop, previous version took 55s, this one takes 14s.

@dfalster which previous version are you referring to? pivot_trait_wider()

**Comment by dfalster** [comment]
Hi Fonti, 

Ooh yes, please do add some tests! Maybe to run on subset of data?

Re previous version - no,  I was referring to the version in the first commit in this PR, so not something you've seen before.

The trait data here are still in long format, but we've merged in info from other tables to create one big table. Having one big table will be preferable in some circumstances (here for an API we are building)

**Comment by fontikar** [comment]
Just waiting for the above tests to past! Adding a few small tests! Is it worth trimming down all the "TO DO:" comments as those will be exported to public?

**Comment by dfalster** [comment]
> Just waiting for the above tests to past! Adding a few small tests! Is it worth trimming down all the "TO DO:" comments as those will be exported to public?

Can we keep for now? These will get actioned in the next few months.

**Comment by fontikar** [comment]
Woohoo tests all passed and merged folks!

**Pull request merged**